src/Compilers/Core/Portable/Serialization/ObjectWriter.cs

```diff
@@ -670,13 +680,15 @@ private void WriteKnownType(Type type)
             this.WriteInt32(_binderSnapshot.GetTypeId(type));
         }
 
-        private void WriteObject(object instance)
+        private void WriteObject(object instance, IObjectWritable instanceAsWritableOpt)
```

Comment:
I tried that (hoping for the exact benefit you mentioend), but didn't measure any improvement.  My guess is that the runtime isn't specializing the dispatch fully since the emitted member is shared among all reference types.  

I know this benefit is huge when you have struct types (since each struct gets its own emitted method, and all the interface methods become direct dispatch), but it looks like it either doesn't happen for reference types, or it is so marginal in perf gain as to not matter :-/